File: 830-PositionsofLargeGroups/830-PositionsofLargeGroups.java
/*

830. Positions of Large Groups
Easy

In a string S of lowercase letters, these letters form consecutive groups of the same character.

For example, a string like S = "abbxxxxzyy" has the groups "a", "bb", "xxxx", "z" and "yy".

Call a group large if it has 3 or more characters.  We would like the starting and ending positions of every large group.

The final answer should be in lexicographic order.
 

Example 1:

Input: "abbxxxxzzy"
Output: [[3,6]]
Explanation: "xxxx" is the single large group with starting  3 and ending positions 6.

Example 2:

Input: "abc"
Output: []
Explanation: We have "a","b" and "c" but no large group.

Example 3:

Input: "abcdddeeeeaabbbcd"
Output: [[3,5],[6,9],[12,14]]


Note:  1 <= S.length <= 1000

*/

class Solution {
    public List<List<Integer>> largeGroupPositions(String S) {
        List<List<Integer>> resList = new ArrayList<>();
        if(!validateInput(S)) return resList;
        int start = 0, end = 1;
        while(end < S.length()){
            while(end < S.length() && S.charAt(start) == S.charAt(end)){
                end++;
                continue;
            }
            if(end - start >= 3){
                List<Integer> list = new ArrayList<>();
                list.add(start);
                list.add(end - 1);
                resList.add(list);
            }
            start = end;
            end = start + 1;
        }
        return resList;
    }
    private boolean validateInput(String S) {
        if(S == null || S.length() == 0) return false;
        return true;
    }
}

File: 830-PositionsofLargeGroups/830-PositionsofLargeGroups.py
"""

830. Positions of Large Groups

In a string s of lowercase letters, these letters form consecutive groups of the same character.

For example, a string like s = "abbxxxxzyy" has the groups "a", "bb", "xxxx", "z", and "yy".

A group is identified by an interval [start, end], where start and end denote the start and end indices (inclusive) of the group. In the above example, "xxxx" has the interval [3,6].

A group is considered large if it has 3 or more characters.

Return the intervals of every large group sorted in increasing order by start index.

Example 1:

Input: s = "abbxxxxzzy"
Output: [[3,6]]
Explanation: "xxxx" is the only large group with start index 3 and end index 6.

Example 2:

Input: s = "abc"
Output: []
Explanation: We have groups "a", "b", and "c", none of which are large groups.

Example 3:

Input: s = "abcdddeeeeaabbbcd"
Output: [[3,5],[6,9],[12,14]]
Explanation: The large groups are "ddd", "eeee", and "bbb".

Constraints:

    1 <= s.length <= 1000
    s contains lowercase English letters only.

"""

class Solution:
    def largeGroupPositions(self, s: str) -> List[List[int]]:
        lo = 0
        res = []
        summ = 0
        mapp = defaultdict(list)
        while lo < len(s) - 1:
            if s[lo] == s[lo + 1]:
                hi = lo
                while hi < len(s) and s[lo] == s[hi]: hi += 1
                if hi - lo >= 3:
                    res.append([lo, hi - 1])
                lo = hi
            else: lo += 1
        return res
